src/PowerShellEditorServices/Workspace/WorkspaceFileSystemWrapper.cs

```diff
@@ -0,0 +1,317 @@
+            _isParentPath = (depth == 0);
+            _factory = factory;
+            _depth = depth;
+        }
+
+        /// <inheritdoc />
+        public override IEnumerable<FileSystemInfoBase> EnumerateFileSystemInfos()
+        {
+            if (!_concreteDirectoryInfo.Exists || _depth >= _factory.MaxRecursionDepth) { yield break; }
+            foreach (FileSystemInfo fileSystemInfo in _factory.SafeEnumerateFileSystemInfos(_concreteDirectoryInfo))
+            {
+                switch (fileSystemInfo)
+                {
+                    case DirectoryInfo dirInfo:
+                        yield return _factory.CreateDirectoryInfoWrapper(dirInfo, _depth + 1);
+                        break;
+                    case FileInfo fileInfo:
+                        yield return _factory.CreateFileInfoWrapper(fileInfo, _depth);
+                        break;
+                    default:
+                        // We should NEVER get here, but if we do just continue on
+                        break;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Returns an instance of <see cref="DirectoryInfoBase" /> that represents a subdirectory.
+        /// </summary>
+        /// <remarks>
+        /// If <paramref name="name" /> equals '..', this returns the parent directory.
+        /// </remarks>
+        /// <param name="name">The directory name.</param>
+        /// <returns>The directory</returns>
+        public override DirectoryInfoBase GetDirectory(string name)
+        {
+            var isParentPath = string.Equals(name, "..", StringComparison.Ordinal);
+
+            if (isParentPath) { return ParentDirectory; }
+
+            var dirs = _concreteDirectoryInfo.GetDirectories(name);
+
+            if (dirs.Length == 1) { return _factory.CreateDirectoryInfoWrapper(dirs[0], _depth + 1); }
+            if (dirs.Length == 0) { return null; }
+            // This shouldn't happen. The parameter name isn't supposed to contain wild card.
+            throw new InvalidOperationException(
+                string.Format(
+                    System.Globalization.CultureInfo.CurrentCulture,
+                    "More than one sub directories are found under {0} with name {1}.",
+                    _concreteDirectoryInfo.FullName, name));
+        }
+
+        /// <inheritdoc />
+        public override FileInfoBase GetFile(string name) => _factory.CreateFileInfoWrapper(new FileInfo(Path.Combine(_concreteDirectoryInfo.FullName, name)), _depth);
+
+        /// <inheritdoc />
+        public override string Name => _isParentPath ? ".." : _concreteDirectoryInfo.Name;
+
+        /// <summary>
+        /// Returns the full path to the directory.
+        /// </summary>
+        public override string FullName => _concreteDirectoryInfo.FullName;
+
+        /// <summary>
+        /// Returns the parent directory.
+        /// </summary>
+        public override DirectoryInfoBase ParentDirectory
```

Comment:
This is relatively involved -- might be better as a method to signal that real work is being done.

Reply:
Maybe.  The "work" is the exception handling, not that actual call.

Reply:
No I mean that it allocates a new object under the hood. Given work we've done to avoid things like that, making this a method is more explicit about the depth of logic under this property, so that callers are more mindful of calling it repeatedly

Reply:
Done.   Also did the same for the File object too.